File: myFinance_improved_V2/lib/features/journal_input/domain/entities/journal_entry.dart
// Domain Entity: JournalEntry
// Pure business entity with no dependencies on frameworks

import 'package:freezed_annotation/freezed_annotation.dart';
import 'transaction_line.dart';

part 'journal_entry.freezed.dart';

@freezed
class JournalEntry with _$JournalEntry {
  const JournalEntry._();

  const factory JournalEntry({
    @Default([]) List<TransactionLine> transactionLines,
    required DateTime entryDate,
    String? overallDescription,
    String? selectedCompanyId,
    String? selectedStoreId,
    String? counterpartyCashLocationId,
  }) = _JournalEntry;

  // Calculated values
  double get totalDebits => transactionLines
      .where((line) => line.isDebit)
      .fold(0.0, (sum, line) => sum + line.amount);

  double get totalCredits => transactionLines
      .where((line) => !line.isDebit)
      .fold(0.0, (sum, line) => sum + line.amount);

  double get difference => totalDebits - totalCredits;

  bool get isBalanced => difference.abs() < 0.01; // Allow for small rounding differences

  int get debitCount => transactionLines.where((line) => line.isDebit).length;

  int get creditCount => transactionLines.where((line) => !line.isDebit).length;

  bool canSubmit() {
    return transactionLines.isNotEmpty &&
           isBalanced &&
           selectedCompanyId != null &&
           selectedCompanyId!.isNotEmpty;
  }

  JournalEntry addTransactionLine(TransactionLine line) {
    return copyWith(
      transactionLines: [...transactionLines, line],
    );
  }

  JournalEntry removeTransactionLine(int index) {
    if (index >= 0 && index < transactionLines.length) {
      final newLines = List<TransactionLine>.from(transactionLines);
      newLines.removeAt(index);
      return copyWith(transactionLines: newLines);
    }
    return this;
  }

  JournalEntry updateTransactionLine(int index, TransactionLine line) {
    if (index >= 0 && index < transactionLines.length) {
      final newLines = List<TransactionLine>.from(transactionLines);
      newLines[index] = line;
      return copyWith(transactionLines: newLines);
    }
    return this;
  }

  JournalEntry clear() {
    return JournalEntry(
      transactionLines: const [],
      entryDate: DateTime.now(),
      overallDescription: null,
      selectedCompanyId: selectedCompanyId, // Keep company ID
      selectedStoreId: selectedStoreId, // Keep store ID
      counterpartyCashLocationId: null,
    );
  }

  // =============================================================================
  // Business Logic Methods (Moved from Presentation Layer)
  // =============================================================================

  /// Calculate suggested amount to balance the journal entry
  /// Returns the absolute difference if entry is unbalanced, null if balanced
  double? getSuggestedAmountForBalance() {
    final diff = difference;
    return diff.abs() > 0.01 ? diff.abs() : null;
  }

  /// Suggest whether the next transaction should be debit or credit
  /// to balance the journal entry
  /// Returns true for debit, false for credit
  bool suggestDebitOrCredit() {
    if (totalDebits < totalCredits) return true;  // Need more debits
    if (totalCredits < totalDebits) return false; // Need more credits
    return true; // Default to debit when balanced
  }

  /// Get all cash location IDs already used in transaction lines
  /// Used to prevent duplicate cash location selection
  Set<String> getUsedCashLocationIds() {
    return transactionLines
        .where((line) => line.categoryTag == 'cash' && line.cashLocationId != null)
        .map((line) => line.cashLocationId!)
        .toSet();
  }
}

File: website/src/features/journal-input/domain/entities/JournalEntry.ts
/**
 * JournalEntry Entity
 * Represents a complete journal entry with multiple transaction lines
 */

import { TransactionLine } from './TransactionLine';

export class JournalEntry {
  constructor(
    public readonly companyId: string,
    public readonly storeId: string | null,
    public readonly date: string,
    public readonly transactionLines: TransactionLine[]
  ) {}

  /**
   * Calculate total debits
   */
  get totalDebits(): number {
    return this.transactionLines
      .filter((line) => line.isDebit)
      .reduce((sum, line) => sum + line.amount, 0);
  }

  /**
   * Calculate total credits
   */
  get totalCredits(): number {
    return this.transactionLines
      .filter((line) => !line.isDebit)
      .reduce((sum, line) => sum + line.amount, 0);
  }

  /**
   * Count debit entries
   */
  get debitCount(): number {
    return this.transactionLines.filter((line) => line.isDebit).length;
  }

  /**
   * Count credit entries
   */
  get creditCount(): number {
    return this.transactionLines.filter((line) => !line.isDebit).length;
  }

  /**
   * Calculate difference (should be 0 for balanced entry)
   */
  get difference(): number {
    return this.totalDebits - this.totalCredits;
  }

  /**
   * Check if journal entry is balanced
   */
  get isBalanced(): boolean {
    return Math.abs(this.difference) < 0.01;
  }

  /**
   * Check if journal entry can be submitted
   */
  canSubmit(): boolean {
    return (
      this.isBalanced &&
      this.transactionLines.length >= 2 &&
      this.transactionLines.every((line) => line.isValid)
    );
  }

  /**
   * Get the cash location ID if any transaction line has one
   * Returns null if no cash location is used
   */
  getCashLocationId(): string | null {
    const lineWithCashLocation = this.transactionLines.find(
      (line) => line.cashLocationId !== null
    );
    return lineWithCashLocation ? lineWithCashLocation.cashLocationId : null;
  }

  /**
   * Check if cash location is already used in this journal entry
   */
  hasCashLocation(): boolean {
    return this.getCashLocationId() !== null;
  }

  /**
   * Add transaction line
   */
  addTransactionLine(line: TransactionLine): JournalEntry {
    return new JournalEntry(
      this.companyId,
      this.storeId,
      this.date,
      [...this.transactionLines, line]
    );
  }

  /**
   * Update transaction line at index
   */
  updateTransactionLine(index: number, line: TransactionLine): JournalEntry {
    const newLines = [...this.transactionLines];
    newLines[index] = line;
    return new JournalEntry(
      this.companyId,
      this.storeId,
      this.date,
      newLines
    );
  }

  /**
   * Remove transaction line at index
   */
  removeTransactionLine(index: number): JournalEntry {
    const newLines = this.transactionLines.filter((_, i) => i !== index);
    return new JournalEntry(
      this.companyId,
      this.storeId,
      this.date,
      newLines
    );
  }
}
